src/Algebra/Graph/Relation/Symmetric.hs

```diff
@@ -11,27 +11,497 @@
+-- hasVertex 1 ('vertex' 2)       == False
+-- hasVertex x . 'removeVertex' x == 'const' False
+-- @
+hasVertex :: Ord a => a -> SymmetricRelation a -> Bool
+hasVertex x = R.hasVertex x . fromSymmetric
+
+-- | Check if a graph contains a given edge.
+-- Complexity: /O(log(n))/ time.
+--
+-- @
+-- hasEdge x y 'empty'            == False
+-- hasEdge x y ('vertex' z)       == False
+-- hasEdge x y ('edge' x y)       == True
+-- hasEdge y x ('edge' x y)       == True
+-- hasEdge x y . 'removeEdge' x y == 'const' False
+-- hasEdge x y                  == 'elem' (x,y) . 'edgeList'
+-- @
+hasEdge :: Ord a => a -> a -> SymmetricRelation a -> Bool
+hasEdge x y = R.hasEdge x y . fromSymmetric
+
+-- | The number of vertices in a graph.
+-- Complexity: /O(1)/ time.
+--
+-- @
+-- vertexCount 'empty'             ==  0
+-- vertexCount ('vertex' x)        ==  1
+-- vertexCount                   ==  'length' . 'vertexList'
+-- vertexCount x \< vertexCount y ==> x \< y
+-- @
+vertexCount :: SymmetricRelation a -> Int
+vertexCount = R.vertexCount . fromSymmetric
+
+-- | The number of edges in a graph.
+-- Complexity: /O(1)/ time.
+--
+-- @
+-- edgeCount 'empty'      == 0
+-- edgeCount ('vertex' x) == 0
+-- edgeCount ('edge' x y) == 2
+-- edgeCount            == 'length' . 'edgeList'
+-- @
+edgeCount :: SymmetricRelation a -> Int
+edgeCount = R.edgeCount . fromSymmetric
+
+-- | The sorted list of vertices of a given graph.
+-- Complexity: /O(n)/ time and memory.
+--
+-- @
+-- vertexList 'empty'      == []
+-- vertexList ('vertex' x) == [x]
+-- vertexList . 'vertices' == 'Data.List.nub' . 'Data.List.sort'
+-- @
+vertexList :: SymmetricRelation a -> [a]
+vertexList = R.vertexList . fromSymmetric
+
+-- | The sorted list of edges of a graph.
+-- Complexity: /O(n + m)/ time and /O(m)/ memory.
+--
+-- @
+-- edgeList 'empty'          == []
+-- edgeList ('vertex' x)     == []
+-- edgeList ('edge' x y)     == [(x,y), (y,x)]
+-- edgeList ('star' 2 [3,1]) == [(2,1), (1,2), (2,3), (3,2)]
+-- @
+edgeList :: SymmetricRelation a -> [(a, a)]
+edgeList = R.edgeList . fromSymmetric
+
+-- | The set of vertices of a given graph.
+-- Complexity: /O(1)/ time.
+--
+-- @
+-- vertexSet 'empty'      == Set.'Set.empty'
+-- vertexSet . 'vertex'   == Set.'Set.singleton'
+-- vertexSet . 'vertices' == Set.'Set.fromList'
+-- @
+vertexSet :: SymmetricRelation a -> Set.Set a
+vertexSet = R.vertexSet . fromSymmetric
+
+-- | The set of edges of a given graph.
+-- Complexity: /O(1)/ time.
+--
+-- @
+-- edgeSet 'empty'      == Set.'Set.empty'
+-- edgeSet ('vertex' x) == Set.'Set.empty'
+-- edgeSet ('edge' x y) == Set.'Set.fromList' [(x,y), (y,x)]
+-- @
+edgeSet :: SymmetricRelation a -> Set.Set (a, a)
+edgeSet = R.edgeSet . fromSymmetric
+
+-- | The sorted /adjacency list/ of a graph.
+-- Complexity: /O(n + m)/ time and /O(m)/ memory.
+--
+-- @
+-- adjacencyList 'empty'          == []
+-- adjacencyList ('vertex' x)     == [(x, [])]
+-- adjacencyList ('edge' 1 2)     == [(1, [2]), (2, [1])]
```

Comment:
Here, I think, it is correct to report the same edge twice, because we are interested in adjacency lists, not individual edges. So, it's fine to say that 1 is adjacent to 2, and that 2 is adjacent to 1. This is what it means for an edge to be undirected.

Reply:
So you're saying that in the Graph data type you dont want to have the same edge twice? how about the performance issues that may arise?

Reply:
Which performance issues are you referring to? It is fairly easy and fast to filter out all edges where the second vertex is greater than the first. For example:

```
edgeList g = filter (uncurry (<=)) (Set.toAscList $ relation g)
```

Does this work?

Reply:
Yes!

Reply:
@snowleopard I'm afraid that doesnt solve for cases like `edge 1 1` where it gives `edgeList (edge 1 1) = []`.
 
I came up with `edgeList g = nubBy (\x y -> x == swap y) (Set.toAscList $ relation g)` but it gets quadratic on the number of edges.